### PR TITLE
Refactor wall-clock benchmarking

### PR DESCRIPTION
**Pull request by mrattle**
### TL;DR - What are you trying to accomplish?

This PR is a refactor of the existing `criterion` benchmarking suite which is used to provide wall-clock benchmarks.  This is one part of augmenting the toolkit that developers and maintainers can use to evaluate proposed code changes.

### Details - How are you making this change?  What are the effects of this change?

Set operations have been rewritten to test against a range of request sizes, and a known distribution generated from production traffic patterns (consistent requests).  This gives a more realistic view of how code changes may skew the time it takes to perform an operation as the payload sizes / size of data on the wire changes.

Sample output from the new benchmarks (click to expand):

<details>

```
     Running benches/request_distribution_bench.rs (target/release/deps/request_distribution_bench-d70fbcdf3f10dcbc)
Gnuplot not found, using plotters backend
set_with_p100_strings   time:   [111.06 µs 111.30 µs 111.60 µs]
                        change: [-11.663% -10.851% -10.050%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  5 (5.00%) high severe

set_with_p99_strings    time:   [52.975 µs 53.032 µs 53.094 µs]
                        change: [-12.813% -11.470% -10.227%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) high mild
  6 (6.00%) high severe

set_with_p95_strings    time:   [38.926 µs 39.029 µs 39.147 µs]
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

set_with_p90_strings    time:   [32.491 µs 32.561 µs 32.628 µs]
Found 16 outliers among 100 measurements (16.00%)
  2 (2.00%) high mild
  14 (14.00%) high severe

set_with_p75_strings    time:   [25.631 µs 25.734 µs 25.844 µs]
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) low severe
  2 (2.00%) high mild
  2 (2.00%) high severe

set_with_p50_strings    time:   [24.315 µs 24.440 µs 24.576 µs]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

set_multi_with_p100_strings
                        time:   [7.6131 ms 7.6360 ms 7.6567 ms]
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) low severe
  4 (4.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe

set_multi_with_p99_strings
                        time:   [2.6297 ms 2.6453 ms 2.6581 ms]
Found 6 outliers among 100 measurements (6.00%)
  6 (6.00%) low severe

Benchmarking set_multi_with_p95_strings: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 6.9s, enable flat sampling, or reduce sample count to 50.
set_multi_with_p95_strings
                        time:   [1.3445 ms 1.3568 ms 1.3672 ms]
Found 13 outliers among 100 measurements (13.00%)
  10 (10.00%) low severe
  1 (1.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe

set_multi_with_p90_strings
                        time:   [851.17 µs 860.47 µs 868.24 µs]
Found 17 outliers among 100 measurements (17.00%)
  13 (13.00%) low severe
  2 (2.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe

set_multi_with_p75_strings
                        time:   [380.73 µs 387.54 µs 395.11 µs]
Found 16 outliers among 100 measurements (16.00%)
  2 (2.00%) low severe
  12 (12.00%) low mild
  2 (2.00%) high severe

set_multi_with_p50_strings
                        time:   [91.974 µs 92.246 µs 92.460 µs]
Found 9 outliers among 100 measurements (9.00%)
  2 (2.00%) low severe
  3 (3.00%) low mild
  2 (2.00%) high mild
  2 (2.00%) high severe

set_multi_with_consistent_requests
                        time:   [468.25 µs 471.83 µs 474.80 µs]
Found 13 outliers among 100 measurements (13.00%)
  5 (5.00%) low severe
  2 (2.00%) low mild
  1 (1.00%) high mild
  5 (5.00%) high severe

add_with_p100_strings   time:   [100.64 µs 101.00 µs 101.29 µs]
Found 14 outliers among 100 measurements (14.00%)
  2 (2.00%) low severe
  5 (5.00%) high mild
  7 (7.00%) high severe

add_with_p99_strings    time:   [53.310 µs 53.480 µs 53.720 µs]
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) high mild
  7 (7.00%) high severe

add_with_p95_strings    time:   [39.382 µs 39.526 µs 39.701 µs]
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) high mild
  6 (6.00%) high severe

add_with_p90_strings    time:   [32.817 µs 33.177 µs 33.739 µs]
Found 9 outliers among 100 measurements (9.00%)
  5 (5.00%) high mild
  4 (4.00%) high severe

add_with_p75_strings    time:   [25.669 µs 25.766 µs 25.875 µs]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

add_with_p50_strings    time:   [24.121 µs 24.232 µs 24.344 µs]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low severe
  1 (1.00%) high mild
  3 (3.00%) high severe

add_multi_with_p100_strings
                        time:   [7.1373 ms 7.1472 ms 7.1561 ms]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild

add_multi_with_p99_strings
                        time:   [2.6607 ms 2.6749 ms 2.6949 ms]
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) low severe
  3 (3.00%) low mild
  3 (3.00%) high severe

Benchmarking add_multi_with_p95_strings: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 6.9s, enable flat sampling, or reduce sample count to 60.
add_multi_with_p95_strings
                        time:   [1.3684 ms 1.3717 ms 1.3745 ms]
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) low severe
  2 (2.00%) low mild
  3 (3.00%) high mild

add_multi_with_p90_strings
                        time:   [841.01 µs 855.63 µs 869.32 µs]
Found 14 outliers among 100 measurements (14.00%)
  11 (11.00%) low severe
  2 (2.00%) low mild
  1 (1.00%) high mild

add_multi_with_p75_strings
                        time:   [390.24 µs 392.23 µs 394.07 µs]
Found 17 outliers among 100 measurements (17.00%)
  8 (8.00%) low severe
  7 (7.00%) low mild
  2 (2.00%) high mild

add_multi_with_p50_strings
                        time:   [89.332 µs 89.686 µs 89.978 µs]
Found 15 outliers among 100 measurements (15.00%)
  3 (3.00%) low severe
  2 (2.00%) low mild
  6 (6.00%) high mild
  4 (4.00%) high severe

add_multi_with_consistent_requests
                        time:   [466.58 µs 472.67 µs 479.91 µs]
Found 9 outliers among 100 measurements (9.00%)
  5 (5.00%) low severe
  1 (1.00%) high mild
  3 (3.00%) high severe
  
```

</details>

### CODE REVIEW

File: benches/bench.rs [hunk 1]
```diff
@@ -342,17 +223,10 @@ fn bench_decrement_no_reply(c: &mut Criterion) {
 criterion_group!(
     benches,
     bench_get,
-    bench_get_many,
+    bench_get_multi,
     bench_get_large,
     bench_get_many_large,
-    bench_set_with_string,
```
[the remark this answers is not in the document]
Comment (mrattle) [reply benches/bench.rs:348]: I'll bring those back in.  I do think that you need to run the full request distribution benchmark to get a more realistic picture of a change, but the simple tests are fine for a very basic comparison.

**Review by danmayer — APPROVED** [review]

**Review by nickamorim — APPROVED** [review]
